Pci 2665 independent chat send add ability to specify sinks put params

Esysc opened the pull request:
Part of: [PCI-2665](https://pix4dbug.atlassian.net/browse/PCI-2665)

This is the first part of the new feature, it's developed around `put.params`. The next PR will address the `source` resource configuration and it will complete the tests. **Note** also that the README will be updated accordingly in a separate PR.

### What is this PR introducing? 
Basically it adds a parameter `sinks` that can be specified in a `put` step. This new parameter is expected as a slice of strings and if nothing is specified it behaviors as before defaulting to all supported sinks. Currently only 2 are supported: `github` and `gchat`.

Example:

`put` step will only send to gchat, no Github status involved. 
```yaml
plan:
      - get: the-repo
        trigger: true
      - put: gchat
        params: 
         state: pending
         sinks: ["gchat"]
      ...
```

Best review commity by commit.

### Todos

- [ ] Update documentation (separate PR)
- [x] Add and update current tests ( next PR)
- [x] Add sinks in source configuration ( again, added in next PR)

Esysc commented on cogito/put.go:
```diff
@@ -16,7 +16,7 @@ type Putter interface {
 	ProcessInputDir() error
 	// Sinks return the list of configured sinks.
-	Sinks() []Sinker
+	Sinks() ([]Sinker, error)
```
`Sinks()`  now return also an error. If for example the sink is misspelled or simply not supported it will be reported.

Esysc commented on cogito/put.go:
```diff
@@ -51,7 +51,11 @@ func Put(log hclog.Logger, input []byte, out io.Writer, args []string, putter Pu
 	var sinkErrors []error
-	for _, sink := range putter.Sinks() {
+	sinks, err := putter.Sinks()
+	if err != nil {
```
Here is where the unsupported sink will raise.

Esysc commented on cogito/put_test.go:
```diff
@@ -249,11 +249,11 @@ func TestPutterProcessInputDirFailure(t *testing.T) {
 
 	testCases := []testCase{
-		{
+		/*{
```
This test is left commented for now, it will be updated in the next PR where the resource Source configuration will be extended.

aliculPix4D replied:
Personally, I would leave it as it is. No comment needed, since the other PR will be merged to this branch before merging to master. Like this it only adds unnecessary changes in git diff.

I understand that if you don't do this, the CI would be red.

Esysc replied:
exactly, another option is to remove this test temporarily.

Esysc commented on cogito/putter.go:
```diff
@@ -65,8 +65,9 @@ func (putter *ProdPutter) LoadConfiguration(input []byte, args []string) error {
 func (putter *ProdPutter) ProcessInputDir() error {
-	// putter.InputDir, corresponding to key "put:inputs:", should contain 1 or 2 dirs.
-	// If it contains one, we support autodiscovery by not requiring to name it, we know
+	// putter.InputDir, corresponding to key "put:inputs:", should contain 0, 1 or 2 dirs.
```
If Cogito put step is configured to only send to chat, it could also have no input dirs at all .

Esysc commented on cogito/putter.go:
```diff
@@ -105,10 +106,10 @@ func (putter *ProdPutter) ProcessInputDir() error {
-		return fmt.Errorf(
-			"put:inputs: missing directory for GitHub repo: have: %v, GitHub: %s/%s",
-			collected, source.Owner, source.Repo)
+		putter.log.Debug("No GitHub repositories in inputs, Cogito will only send to chat")
```
If there are 0 directories, Cogito will send only to chat. Any configuration issue will be intercepted earlier (tuned in next PR).

Esysc commented on cogito/putter.go:
```diff
@@ -134,22 +135,40 @@ func (putter *ProdPutter) ProcessInputDir() error {
 		},
 	}
+	sinksParams := putter.Request.Params.Sinks
+	if len(sinksParams) == 0 {
```
Let's default to all supported sinks if nothing is specified.

odormond commented on cogito/putter.go:
```diff
@@ -134,22 +135,40 @@ func (putter *ProdPutter) ProcessInputDir() error {
 	}
+	sinksParams := putter.Request.Params.Sinks
+	if len(sinksParams) == 0 {
+		// No sink specified, it will return all supported sinks.
+		sinksParams = make([]string, 0, len(supportedSinks))
+		for k := range supportedSinks {
+			sinksParams = append(sinksParams, k)
+		}
```
While this is currently true, it might not stay as such when adding another sink. In my understanding, the current behaviour is only here to provide backward compatibility. So this should be changed to something like:
```suggestion
		// No sink specified, we default to github and ghcat for backward compatibility.
		sinksParams = []string{"github", "gchat"}
```

odormond commented on cogito/putter.go:
```diff
@@ -134,22 +135,40 @@ func (putter *ProdPutter) ProcessInputDir() error {
+	}
+	sinkers := make([]Sinker, 0, len(sinksParams))
+	for _, s := range sinksParams {
+		sinkers = append(sinkers, supportedSinks[s])
```
What will happen if `s` is not in `supportedSinks`?

Esysc replied:
It's tested here: https://github.com/Pix4D/cogito/pull/109/files#diff-4642f345de08ea53e3e6fe15f6b457e7052f316cda64aca54e21f78f51204cb3R203  (the other PR). The  cinfured sink is `pizza` and the error `put: source: invalid sink: [pizza]. Supported sinks: 'gchat', 'github'`

odormond commented on cogito/putter.go:
```diff
@@ -134,22 +135,40 @@ func (putter *ProdPutter) ProcessInputDir() error {
+		sinkers = append(sinkers, supportedSinks[s])
+	}
+
+	if len(sinkers) == 0 {
```
This could only be true if it did not blow up above with unknown sinks. In that case, you could as well have encountered a config like `sinks: [github, slack]` and so `sinkers` would hold only the `github` sink and you would silently ignore the `slack` sink. Or it blew up above and that test is useless because you ensure that `sinkParams` is not empty.

Esysc replied:
Good point, thanks

Esysc replied:
This should be addressed in https://github.com/Pix4D/cogito/pull/108/commits/0984de9d8541a7b939fea573f891aaf4c998c457

Esysc commented:
Notes to reviewers:
- This PR is completed by https://github.com/Pix4D/cogito/pull/109
- Status check is still red, I didn't update the GitHub PAT yet.

marco-m-pix4d commented:
Could you please update the PAT and rerun the CI for both PRs?
We cannot merge with failing tests, and without passing tests we don't know what to review, the code could be proved wrong by the tests themselves. Merci.

Esysc commented:
> Could you please update the PAT and rerun the CI for both PRs? We cannot merge with failing tests, and without passing tests we don't know what to review, the code could be proved wrong by the tests themselves. Merci.

Done!

aliculPix4D commented on cogito/protocol_test.go:
```diff
@@ -209,6 +209,7 @@ func TestPutParamsPrintLogRedaction(t *testing.T) {
 		ChatMessage:     "stecchino",
 		ChatMessageFile: "dir/msg.txt",
 		GChatWebHook:    "sensitive-gchat-webhook",
+		Sinks:           []string{"gchat", "github"},
```
you can as well use some random names here. To be consistent with `johnny` and `stecchino`

Esysc replied:
That one should be real, I could also remove it because it defaults to all supported sinks.

aliculPix4D replied:
Can you explain why? I am missing this point. Is `func (putter *ProdPutter) Sinks() ([]Sinker, error)` somehow called before?

My understanding only String method is called in this test.

aliculPix4D replied:
Ah it is called: `params.Sinks`

aliculPix4D commented on cogito/put.go:
```diff
@@ -16,7 +16,7 @@ type Putter interface {
 	// ProcessInputDir validates and extract the needed information from the "put input".
 	ProcessInputDir() error
 	// Sinks return the list of configured sinks.
```
and a error. Either change the comment to includes returns an error, or write the comment differently to say what the method does.

Esysc replied:
I have updated the comment, `Sinks()` validate also if a specific sink is supported.

marco-m-pix4d commented:
Hello @Esysc, could you please mention here (and in the other PR) when all the outstanding comments have been answered, so that I can have an overview look? (no pressure, take your time!). Merci :-)

aliculPix4D commented on cogito/put_test.go:
```diff
@@ -309,7 +309,7 @@ func TestPutterProcessInputDirNonExisting(t *testing.T) {
 	putter := cogito.NewPutter("dummy-API", hclog.NewNullLogger())
 
-	sinks := putter.Sinks()
+	sinks, _ := putter.Sinks()
```
we should check that error is nil first.

Esysc replied:
thanks, I addressed your point.

aliculPix4D commented on cogito/put_test.go:
```diff
@@ -309,8 +309,8 @@ func TestPutterProcessInputDirNonExisting(t *testing.T) {
-	sinks := putter.Sinks()
-
+	sinks, err := putter.Sinks()
+	assert.Assert(t, err == nil)
```
Or maybe this is better?: https://github.com/Pix4D/fly_helper/blob/15f99bdfd4f5462f4f7be321e8072c46cc346bac/pkg/pipelinesetter/pipelinesetter_test.go#L178

Esysc replied:
> LGTM. I left some minor comments.
> 
> I don't like the commented test. So its a must that you merge the other PR into this one before, merging both to master.

Me neither, however I will merge the other PR here, this was already planned to not merge independently.

Esysc replied:
> Or maybe this is better?: https://github.com/Pix4D/fly_helper/blob/15f99bdfd4f5462f4f7be321e8072c46cc346bac/pkg/pipelinesetter/pipelinesetter_test.go#L178

Yes, I'll do :)

Esysc replied:
Fixed

aliculPix4D reviewed (approved):
LGTM. I left some minor comments. 

I don't like the commented test. So its a must that you merge the other PR into this one before, merging both to master.

Esysc commented:
@Pix4D/integration , this PR is ready, I've addressed all the feedback I received until now. Note that it will be not merged as is, but after merging #109 that will add all the missing features and the test coverage.

Esysc commented:
The CI is failing because of  test commit SHA reached the max statuses, however this is fixed in the next PR that will be merge onto this branch. For this reason I'd like to keep as is showing that this PR is not intended to be merged until the other one will be.

marco-m-pix4d commented:
Rebased on #110, this will fix the CI

Esysc commented:
Closing, replaced by https://github.com/Pix4D/cogito/pull/113
